### DISCUSSION TITLE
Libvips seems to keep memory even after releasing

### DISCUSSION THREAD

**Discussion by bnbourgade** [libvips/libvips#4184] [Q&A]
## Bug report

**Describe the bug**
I am using libvips (with pyvips) to infer over a list of pdf files in order to convert their pages into JPG files, and I noticed that the RAM consumption of the code was increasing a lot while running, but was not released unless the container was removed, showing a possible memory leak from vips.

The benchmarking i've been performing to measure the ram consumption provided me the following results :
![CEV_mem_usage](https://github.com/user-attachments/assets/3c707138-5f53-4dfb-9540-403d1eefbba4)

**To Reproduce**

The code I have been using for infering on the list of pdf is the following :

```python
from pyvips import Image
from pathlib import Path

input_dir = Path(...) # Points to a folder that contains only pdf files
output_path = Path(...) # Points to the folder we want to save the extracted jpg

for file_name in input_folder.iterdir():
    img = Image.new_from_file(str(file_name), access="sequential", page=0, dpi=200)
    img.write_to_file(f"{output_path}/{file_name.stem}.jpg")
```

**Expected behavior**
As we can see on the left of the curve, there is a little jump of the RAM consumption, dut to me testing my container with 1 PDF file to extract only. In a normal case, we should have a rise of RAM consumtion, then after it should go back to normal. Same when running many extractions, after the end of all of them the RAM consumption should go back to its level before starting inferences.

**Actual behavior**
Theres is a constant rise of RAM consumption while extracting PDF files, and after the extraction is finished, the ram consumption stays at the level it was after the end of the last file's extraction.

**Environment**
Docker image based on Ubuntu 22.04
libvips=8.12.1build1
pyvips=2.2.3

**Comment by jcupitt** [comment]
Hello @bnbourgade,

There are probably a few issues here:

### Operation cache

libvips keeps a cache of recently opened images, so that if you open an image a second time, it's very fast. If you are iterating over a directory of different images, ie. no reuse, you don't need this feature. You can disable it with:

```python
pyvips.cache_set_max(0)
```

meaning set the number of operations to cache to zero. There are other cache controls, see:

https://libvips.github.io/pyvips/voperation.html?highlight=cache#pyvips.voperation.cache_set_max

### Memory fragmentation

Even with the operation cache disabled, on some platforms you will see process RSS creep up over time due to heap fragmentation. You typically see a curve like the one in your graph.

The fix is to switch to a malloc which prioritises low fragmentation, such as jemalloc:

https://jemalloc.net

Though first I'd test just disabling the operation cache. 

### Old libvips

Your libvips is rather old, memory behaviour may have improved in a more recent version.

**Reply by bnbourgade** [reply]
Hello @jcupitt ,

Thanks for your answer, but i need to precise a few things about the solutions you provided.

-  For the Operation Cache solution, i've tried this solution, but to no avail, the RAM consumption was still having the same behavior as if there was no cache limitation
- For the jemalloc solution, I don't know about that, so i will dig into it and provide feedback if necessary
- Finally, for the libvips version, this version was the one provided when installing with `apt-get install`, so I didn't really choose this version in particular. Do you know any way I could install a newer version of the lib ?

**Comment by jcupitt** [comment]
Ah you're right, the `sequential` will stop caching of your PDF.

I tried this test program on Ubuntu 24.04 with libvips 8.16 and python 3.12:

```python
#!/usr/bin/env python3

import pyvips

for i in range(1, 10000):
    print(f"image {i} ...")
    image = pyvips.Image.new_from_file(f"sample/{i}.pdf", access="sequential", page=0, dpi=200)
    image.write_to_file("x.jpg")
```

With this test data:

```
$ mkdir sample
$ for i in {1..10000}; do cp ~/pics/nipguide.pdf sample/$i.pdf; done
```

For this test PDF (1.5mb):

http://www.rollthepotato.net/~john/nipguide.pdf

And watched RSS in `top`. Running:

```
$ ./page0.py
loop 1 ...
loop 2 ...
...
```

I saw:

loop | RSS (mb)
---- | ----
50 | 400
100 | 600
500 | 663
1000 | 668
2000 | 675
4000 | 735
8000 | 736

Which I suppose is roughly the curve you saw, is that right? 

I ran with `jemalloc`:

```
$ LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2 ./page0.py
loop 1 ...
...
```

And watched RSS again:

loop | RSS (mb)
---- | ----
50 | 190
100 | 197
500 | 197
1000 | 188
2000 | 193

Now memory use is low and pretty stable, and will occasionally even drop.

> In a normal case, we should have a rise of RAM consumtion, then after it should go back to normal. Same when running many extractions, after the end of all of them the RAM consumption should go back to its level before starting inferences.

Memory doesn't usually work like this. Libraries like libpoppler, the thing libvips uses for PDF rendering, use the Unix malloc heap for memory allocation. This starts at low addresses and grows upward when necessary using the `brk()` syscall to get more pages of memory from the host OS. 

https://stackoverflow.com/questions/6988487/what-does-the-brk-system-call-do

When programs free memory, the malloc implementation will return pages to the host with `sbrk()`, but it can only do this for complete, unused pages at the end of the heap segment. If the program allocates 100mb, then one byte, then frees the 100mb, the 100mb cannot be returned to the OS until the one byte allocation is freed.

It gets even worse for threaded programs (libvips is highly threaded). malloc keeps a separate small heap for each thread, plus a large global heap. Now fragmentation is horribly complicated, with threads starting and stopping and blocking each other. 

The default malloc implementation in glibc aims to be quick, it does not aim to prevent fragmentation in long-running, highly-threaded server programs. jemalloc (there are several others, but jemalloc is the best-known) uses many tricks to reduce fragmentation, and will return memory to the host when possible. It is (usually) very slightly slower.

**Comment by bnbourgade** [comment]
The usage of `jemalloc` helped me solve my problem. By setting up the environment variable `LD_PRELOAD` and running my code (including the cache max set to 0), I managed to have a RAM consumption over time fitting more what I was expecting in the first place :
![Figure_2](https://github.com/user-attachments/assets/cf21d424-e19d-4e8c-960c-6c3882da8164)

I wrote `ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2` in my Dockerfile.

Thanks for your help !

**Reply by jcupitt** [reply]
Great!

You probably don't need to turn the cache off, fwiw.

**Comment by miroljub1995** [comment]
@jcupitt thank you for the explanation. We are a little concerned about setting `LD_PRELOAD` because it could affect the whole application (dotnet binding), not just `libvips`.

I'm corious what bytes are left at the end of the heap after we finish with image processing so the whole last block could not be released?

**Reply by jcupitt** [reply]
Heap fragmentation is very complex in threaded programs. Each time a thread starts, a new 2mb (usually) block is allocated for the stack. Additionally, glibc malloc will allocate a private heap for the thread (usually called an arena) which can be used for thread malloc without needing to lock the main heap. If a thread at the top of the heap stays alive while a thread lower down stops, a lot of the memory for the lower thread can't be returned to the OS.

I would set `LD_PRELOAD` and test it, it should be fine. Facebook (for example) change the default memory allocator to jemalloc in this way, it is a very common technique.

You could also look at mimalloc, the Microsoft equivalent:

https://github.com/microsoft/mimalloc

**Reply by kleisauke** [reply]
As a possible alternative when using the default Linux glibc memory allocator, you can set the [`MALLOC_ARENA_MAX`](https://sourceware.org/glibc/manual/latest/html_node/Memory-Allocation-Tunables.html) environment variable. For context, see the Ruby-focused blog post [here](https://www.mikeperham.com/2018/04/25/taming-rails-memory-bloat/) and the [in-depth analysis it links to](https://www.speedshop.co/2017/12/04/malloc-doubles-ruby-memory.html).
```env
export MALLOC_ARENA_MAX="2"
```

**Reply by miroljub1995** [reply]
Do you know if `musl libc` has similar fragmentation problems like `glibc`?

**Reply by jcupitt** [reply]
musl libc is much better against fragmentation, but has some other issues which may affect stability (eg. a tiny tiny tiny per thread stack).

**Reply by miroljub1995** [reply]
Setting `MALLOC_ARENA_MAX` to the number of cores indeed helped. We will monitor long term and see if some drawbacks come out.
Thank you all.
